src/sagas/walletConnect.js

```diff
@@ -0,0 +1,424 @@
+  }));
+
+  const { accept } = yield race({
+    accept: take(onAcceptAction.type),
+    reject: take(onRejectAction.type),
+  });
+
+  if (!accept) {
+    yield call(() => web3wallet.rejectSession({
+      id: params.id,
+      reason: {
+        code: USER_REJECTED,
+        message: 'User rejected the session',
+      },
+    }));
+  }
+
+  try {
+    yield call(() => web3wallet.approveSession({
+      id,
+      relayProtocol: params.relays[0].protocol,
+      namespaces: {
+        hathor: {
+          accounts: [`hathor:${NETWORK}:${firstAddress}`],
+          chains: [`hathor:${NETWORK}`],
+          events: AVAILABLE_EVENTS,
+          methods: AVAILABLE_METHODS,
+        },
+      },
+    }));
+
+    yield call(refreshActiveSessions);
+  } catch(error) {
+    console.error('Error on sessionProposal: ', error);
+    yield put(onExceptionCaptured(error));
+  }
+}
+
+export function* onQrCodeRead(action) {
+  const { web3wallet, core } = yield select((state) => state.walletConnect);
+
+  if (!web3wallet) {
+    throw new Error('Wallet connect instance is new and QRCode was read');
+  }
+
+  const { payload } = action;
+
+  try {
+    yield call(() => core.pairing.pair({ uri: payload }));
+  } catch(error) {
+    console.error('Error pairing with QrCode: ', error); 
+    yield put(onExceptionCaptured(error));
+  }
+}
+
+export function* featureToggleUpdateListener() {
+  while (true) {
+    const oldWalletConnectEnabled = yield call(isWalletConnectEnabled);
+    yield take('FEATURE_TOGGLE_UPDATED');
+    const newWalletConnectEnabled = yield call(isWalletConnectEnabled);
+
+    if (oldWalletConnectEnabled && !newWalletConnectEnabled) {
+      yield put({ type: 'WC_SHUTDOWN' });
+    }
+  }
+}
+
+export function* onCancelSession(action) {
+  const { web3wallet } = yield select((state) => state.walletConnect);
+
+  const activeSessions = yield call(() => web3wallet.getActiveSessions());
+
+  if (!activeSessions[action.payload]) {
+    return;
+  }
+
+  yield call(() => web3wallet.disconnectSession({
+    topic: activeSessions[action.payload].topic, 
+    reason: {
+      code: USER_DISCONNECTED,
+      message: 'User cancelled the session',
+    },
+  }));
+
+  yield call(refreshActiveSessions);
+}
+
+/**
+ * This event can be triggered by either the wallet or dapp, indicating the
+ * termination of a session. Emitted only after the session has been
+ * successfully deleted.
+ */
+export function* onSessionDelete(action) {
+  yield call(onCancelSession, action);
+}
+
+export function* saga() {
+  yield all([
+    takeLatest(types.START_WALLET_SUCCESS, init),
+    takeEvery('WC_SESSION_REQUEST', onSessionRequest),
+    takeEvery('WC_SESSION_PROPOSAL', onSessionProposal),
+    takeEvery('WC_SESSION_DELETE', onSessionDelete),
+    takeEvery('WC_CANCEL_SESSION', onCancelSession),
+    takeEvery(types.RESET_WALLET, onWalletReset),
+    takeLatest('SIGN_MESSAGE_REQUEST', onSignMessageRequest),
+    takeLatest(types.WC_QRCODE_READ, onQrCodeRead),
```

Comment:
Why take latest and not every?

Reply:
To prevent errors when the user pressed the button twice on manual connection